src/imputation/tmi_imputation.py

```diff
@@ -0,0 +1,377 @@
+
+    # Replace Nan with zero for companies with NO R&D Q604 = "No"
+    no_rd = filter_by_column_content(df, "604", ["No"])
+
+    for i in target_variables:
+        no_rd[i] = fill_zeros(no_rd, i)
+
+    df = apply_to_original(no_rd, df)
+
+    # Return cleaned original dataset
+    return df
+
+
+def tmi_pre_processing(df, target_variables_list: list) -> pd.DataFrame:
+    """Function that brings together the steps needed before calculating
+    the trimmed mean"""
+    # Create a copy for reference
+    copy_df = df.copy()
+
+    # Filter for long form data
+    long_df = filter_by_column_content(copy_df, "formtype", ["0001"])
+
+    # Filter for instance is not 0
+    filtered_df = long_df.loc[long_df["instance"] != 0]
+
+    # Filter for clear statuses
+    clear_statuses = ["210", "211"]
+    clear_df = filter_by_column_content(filtered_df, "statusencoded", clear_statuses)
+
+    # Fill zeros for no r&d and apply to original
+    df = apply_fill_zeros(clear_df, df, target_variables_list)
+
+    # Calculate imputation classes for each row
+    imp_df = create_imp_class_col(df, "200", "201", "imp_class")
+
+    return imp_df
+
+
+def sort(target_variable: str, df: pd.DataFrame) -> pd.DataFrame:
+    "Sorts a dataframe using the list and order provided."
+    sort_list = [
+        f"{target_variable}",
+        "employees",
+        "reference",
+    ]
+    sorted_df = df.sort_values(
+        by=sort_list,
+        ascending=[True, False, True],
+    )
+
+    return sorted_df
+
+
+def trim_check(
+    df: pd.DataFrame, check_value=10
+) -> pd.DataFrame:  # TODO add check_value to a config
+    """Checks if the number of records is above or below
+    the threshold required for trimming.
+    """
+    # tag for those classes with more than check_value (currently 10)
+    if len(df) <= check_value:  # TODO or is this just <
+        df["trim_check"] = "below_trim_threshold"
+    else:
+        df["trim_check"] = "above_trim_threshold"
+
+    return df
+
+
+def trim_bounds(
+    df: pd.DataFrame,
+    variable: str,
+    lower_perc=15,  # TODO add percentages to config -
+    # check method inBERD_imputation_spec_V3
+    upper_perc=15,
+) -> pd.DataFrame:
+    """Applies a marker to specifiy if a value is to be trimmed or not.
+    NOTE: Trims only if more than 10 valid records
+    """
+
+    # Save the index before the sorting
+    df["pre_index"] = df.index
+
+    # Filter out zero values for trim calculations
+    df = df.loc[df[variable] != 0]
+
+    if len(df) <= 10:
+        df[f"{variable}_trim"] = "dont trim"
+    else:
+        df = filter_by_column_content(df, "trim_check", ["above_trim_threshold"])
+        df.reset_index(drop=True, inplace=True)
+
+        # define the bounds for trimming
+        remove_lower = np.ceil(len(df) * (lower_perc / 100))
+        remove_upper = np.ceil(len(df) * (1 - upper_perc / 100))
+
+        # create trim tag (distinct from trim_check)
+        # to mark which to trim for mean growth ratio
```

Comment:
# To match the SML, we trim for rows less than remove_lower and greater than remove_upper